### PR TITLE
Suspend ASG actions during deployment

### PR DESCRIPTION
**Pull request by grahamar**
- Move de-register & docker stop into BeforeInstall hook as ApplicationStop hook runs with the scripts from the previous successfull install.
- Suspend AZRebalance, ScheduledActions & AlarmNotification ASG actions during a deployment, to prevent new instances coming up or instances being destroyed during deployment.

### CODE REVIEW

**Comment by k4sbasia** [comment]
image

File: nova/core/templates/deploy-scripts/common_functions.sh [hunk 1]
```diff
@@ -101,6 +101,26 @@ autoscaling_enter_standby() {
         return 0
     fi
 
+    msg "Checking to see if ASG ${asg_name} had AZRebalance process suspended previously"
+    local azrebalance_suspended=$($AWS_CLI autoscaling describe-auto-scaling-groups \
+        --auto-scaling-group-name "${asg_name}" \
+        --query 'AutoScalingGroups[].SuspendedProcesses' \
+        --output text | grep -c 'AZRebalance')
```
Comment (gheine) [inline nova/core/templates/deploy-scripts/common_functions.sh]: this should check for ScheduledActions and AlarmNotification too, as they are disabled below too

**Comment by gheine** [comment]
👎   Instead of bunching in ScheduledActions and AlarmNotification together with AZRebalance, the script should similarly check, suspend/flag and resume the

**Comment by RoryH** [comment]
This is awesome!

**Comment by grahamar** [comment]
@gheine I'm not sure I agree, it feels verbose for the sake of being verbose. The CLI provides a way to suspend a list of actions, I'd rather change the naming of things to be generic across auto-scaling actions. I'll definitely add the check to the grep though...

**Pull request merged**